// topology-core/renderLayer.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var le5le_store_1 = require("le5le-store");
var canvas_1 = require("./canvas");
var RenderLayer = /** @class */ (function (_super) {
    __extends(RenderLayer, _super);
    function RenderLayer(parentElem, options) {
        if (options === void 0) { options = {}; }
        var _this = _super.call(this, parentElem, options) || this;
        _this.parentElem = parentElem;
        _this.options = options;
        _this.offscreen = le5le_store_1.Store.get('LT:offscreen');
        _this.parentElem.appendChild(_this.canvas);
        if (_this.data.bkImage) {
            _this.bkImg = new Image();
            _this.bkImg.src = _this.data.bkImage;
            _this.bkImg.onload = function () {
                _this.bkImgRect = _this.coverRect(_this.canvas.width, _this.canvas.height, _this.bkImg.width, _this.bkImg.height);
                _this.render();
            };
        }
        return _this;
    }
    RenderLayer.prototype.render = function () {
        var ctx = this.canvas.getContext('2d');
        ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
        if (this.bkImg) {
            ctx.drawImage(this.bkImg, this.bkImgRect.x, this.bkImgRect.y, this.bkImgRect.width, this.bkImgRect.height, 0, 0, this.width, this.height);
        }
        if (this.data.bkColor) {
            ctx.fillStyle = this.data.bkColor;
            ctx.fillRect(0, 0, this.width, this.height);
        }
        ctx.drawImage(this.offscreen, 0, 0, this.width, this.height);
    };
    RenderLayer.prototype.coverRect = function (canvasWidth, canvasHeight, imgWidth, imgHeight) {
        var x = 0;
        var y = 0;
        var width = imgWidth;
        var height = imgHeight;
        if (imgWidth > imgHeight || (imgWidth === imgHeight && canvasWidth < canvasHeight)) {
            width = canvasWidth * height / canvasHeight;
            x = (imgWidth - width) / 2;
        }
        else if (imgWidth < imgHeight || (imgWidth === imgHeight && canvasWidth > canvasHeight)) {
            height = canvasHeight * width / canvasWidth;
            y = (imgHeight - height) / 2;
        }
        return {
            x: x,
            y: y,
            width: width,
            height: height
        };
    };
    return RenderLayer;
}(canvas_1.Canvas));
exports.RenderLayer = RenderLayer;
